metals/src/main/scala/scala/meta/internal/metals/InteractiveSemanticdbs.scala

```diff
@@ -0,0 +1,196 @@
+package scala.meta.internal.metals
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+import ch.epfl.scala.bsp4j.ScalacOptionsItem
+import java.net.URI
+import java.nio.charset.Charset
+import java.nio.file.Files
+import java.nio.file.Paths
+import java.util.Collections
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicReference
+import org.eclipse.lsp4j.DiagnosticSeverity
+import org.eclipse.lsp4j.PublishDiagnosticsParams
+import org.eclipse.{lsp4j => l}
+import scala.concurrent.ExecutionContext
+import scala.meta.interactive.InteractiveSemanticdb
+import scala.meta.internal.io.FileIO
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.mtags.MtagsEnrichments._
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.internal.mtags.TextDocumentLookup
+import scala.meta.internal.tokenizers.PlatformTokenizerCache
+import scala.meta.internal.{semanticdb => s}
+import scala.meta.io.AbsolutePath
+import scala.tools.nsc.interactive.Global
+
+/**
+ * Produces SemanticDBs on-demand by using the presentation compiler.
+ *
+ * Only used to provide navigation inside external library sources, not used to compile
+ * workspace sources.
+ *
+ * Uses persistent storage to keep track of what external source file is associated
+ * with what build target (to determine classpath and compiler options).
+ */
+final class InteractiveSemanticdbs(
+    workspace: AbsolutePath,
+    buildTargets: BuildTargets,
+    charset: Charset,
+    client: MetalsLanguageClient,
+    tables: Tables,
+    messages: Messages,
+    statusBar: StatusBar
+)(implicit ec: ExecutionContext)
+    extends Cancelable
+    with Semanticdbs {
+  import messages._
+  private val activeDocument = new AtomicReference[Option[String]](None)
+  private val globalCache = Collections.synchronizedMap(
+    new java.util.HashMap[BuildTargetIdentifier, Global]
+  )
+  private val textDocumentCache = Collections.synchronizedMap(
+    new java.util.HashMap[AbsolutePath, s.TextDocument]
+  )
```

Comment:
Have you considered the trade-offs wrt `ConcurrentHashMap`? I'm asking because `ConcurrentHashMap` is my go-to for these scenarios since it's generally faster.

Reply:
I didn't use `ConcurrentHashMap` since the `computeIfAbsent` instructions say the compute function should be "short and simple" and in this class it's expensive. However, worst case it blocks with synchronized either way so can just use `ConcurrentHashMap`. I agree that's better for consistency 👍

Reply:
Turns out that `ConcurrentHashMap` results in a deadlock or infinite loop at

```
"ForkJoinPool.commonPool-worker-1" #96 daemon prio=5 os_prio=31 tid=0x00007f9cec46c000 nid=0x7323 runnable [0x0000700008402000]
   java.lang.Thread.State: RUNNABLE
        at java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1069)
        at java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006)
        at scala.meta.internal.metals.InteractiveSemanticdbs.$anonfun$compile$3(InteractiveSemanticdbs.scala:98)
```

Back to `Collections.synchronizedMap` it is then.